SwiftCompilerSources/Sources/Optimizer/Utilities/EscapeInfoWalker.swift

```diff
@@ -0,0 +1,834 @@
+      } else {
+        return walkUp(value: value, path: path, state: state)
+      }
+    }
+    return false
+  }
+  
+  /// Returns true if the selected address(es) of `lhs`/`lhsPath` can reference the same field as
+  /// the selected address(es) of `rhs`/`rhsPath`.
+  ///
+  /// Example:
+  ///   %1 = struct_element_addr %s, #field1    // true for (%1, %s)
+  ///   %2 = struct_element_addr %s, #field2    // true for (%2, %s), false for (%1,%2)
+  ///
+  mutating func canReferenceSameField(_ lhs: Value, path lhsPath: Path = Path(.anyValueFields),
+                                      _ rhs: Value, path rhsPath: Path = Path(.anyValueFields)) -> Bool {
+    
+    struct Visitor : EscapeInfoWalkerVisitor {
+      let target: Value
+      func visitUse(operand: Operand, path: Path, state: State) -> UseVisitResult {
+        // Note: since we are checking the value of an operand, we are ignoring address
+        // projections with no uses. This is no problem. It just requires a fix_lifetime for
+        // each address to test in alias-analysis test files.
+        if operand.value == target { return .abort }
+        if operand.instruction is ReturnInst { return .ignore }
+        return .continueWalk
+      }
+    }
+    
+    // lhs -> rhs will succeed (= return false) if lhs is a non-escaping "local" object,
+    // but not necessarily rhs.
+    if !isEscaping(addressesOf: lhs, path: lhsPath, visitor: Visitor(target: rhs)) {
+      return false
+    }
+    // The other way round: rhs -> lhs will succeed if rhs is a non-escaping "local" object,
+    // but not necessarily lhs.
+    if !isEscaping(addressesOf: rhs, path: rhsPath, visitor: Visitor(target: lhs)) {
+      return false
+    }
+    return true
+  }
+  
+  mutating func walkDown(value: Operand, path: Path, state: State) -> Bool {
+    // FIXME: match on PartialApply
+    if hasRelevantType(value.value, at: path) || value.value is PartialApplyInst {
+      if DEBUGWALK {
+        print("visitUse \"\(path)\": #\(value.index) \(value.instruction)")
+      }
+      switch visitor.visitUse(operand: value, path: path, state: state) {
+      case .continueWalk:
+        return walkDownDefault(value: value, path: path, state: state)
+      case .ignore:
+        return false
+      case .abort:
+        return true
+      }
+    }
+    return false
+  }
+  
+  mutating func walkDown(address: Operand, path: Path, state: State) -> Bool {
+    if hasRelevantType(address.value, at: path) {
+      if DEBUGWALK {
+        print("visitUse \"\(path)\": #\(address.index) \(address.instruction)")
+      }
+      switch visitor.visitUse(operand: address, path: path, state: state) {
+      case .continueWalk:
+        return walkDownDefault(address: address, path: path, state: state)
+      case .ignore:
+        return false
+      case .abort:
+        return true
+      }
+    }
+    return false
+  }
+  
+  mutating func leafUse(any operand: Operand, path: Path, state: State) -> Bool {
```

Comment:
Why not handling the value/address instructions directly in the dedicated `leaveUse(value:)`/`leafUse(address:)` functions?